File: deepeval/metrics/summarization/summarization.py
from typing import List, Optional, Union
import asyncio

from deepeval.test_case import (
    LLMTestCase,
    LLMTestCaseParams,
    ConversationalTestCase,
)
from deepeval.metrics import BaseMetric
from deepeval.models import DeepEvalBaseLLM
from deepeval.utils import get_or_create_event_loop, prettify_list
from deepeval.metrics.utils import (
    construct_verbose_logs,
    trimAndLoadJson,
    check_llm_test_case_params,
    initialize_model,
)
from deepeval.metrics.summarization.template import SummarizationTemplate
from deepeval.metrics.faithfulness.template import FaithfulnessTemplate
from deepeval.metrics.indicator import metric_progress_indicator
from deepeval.metrics.summarization.schema import *
from deepeval.metrics.faithfulness.schema import *


class SummarizationMetric(BaseMetric):

    _required_params: List[LLMTestCaseParams] = [
        LLMTestCaseParams.INPUT,
        LLMTestCaseParams.ACTUAL_OUTPUT,
    ]

    def __init__(
        self,
        threshold: float = 0.5,
        n: int = 5,
        model: Optional[Union[str, DeepEvalBaseLLM]] = None,
        assessment_questions: Optional[List[str]] = None,
        include_reason: bool = True,
        async_mode=True,
        strict_mode: bool = False,
        verbose_mode: bool = False,
        truths_extraction_limit: Optional[int] = None,
    ):
        self.threshold = 1 if strict_mode else threshold
        self.model, self.using_native_model = initialize_model(model)
        self.evaluation_model = self.model.get_model_name()

        if assessment_questions is not None and len(assessment_questions) == 0:
            self.assessment_questions = None
        else:
            self.assessment_questions = assessment_questions

        self.include_reason = include_reason
        self.n = n
        self.async_mode = async_mode
        self.strict_mode = strict_mode
        self.verbose_mode = verbose_mode

        self.truths_extraction_limit = truths_extraction_limit
        if self.truths_extraction_limit is not None:
            self.truths_extraction_limit = max(self.truths_extraction_limit, 0)

    def measure(
        self,
        test_case: LLMTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
    ) -> float:

        check_llm_test_case_params(test_case, self._required_params, self)

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self, _show_indicator=_show_indicator, _in_component=_in_component
        ):
            if self.async_mode:
                loop = get_or_create_event_loop()
                loop.run_until_complete(
                    self.a_measure(
                        test_case,
                        _show_indicator=False,
                        _in_component=_in_component,
                    )
                )
            else:
                self.truths: List[str] = self._generate_truths(test_case.input)
                self.claims: List[str] = self._generate_claims(
                    test_case.actual_output
                )
                self.coverage_verdicts: List[SummarizationCoverageVerdict] = (
                    self._generate_coverage_verdicts(test_case)
                )
                self.alignment_verdicts: List[SummarizationAlignmentVerdict] = (
                    self._generate_alignment_verdicts()
                )
                alignment_score = self._calculate_score(ScoreType.ALIGNMENT)
                coverage_score = self._calculate_score(ScoreType.COVERAGE)
                self.score_breakdown = {
                    ScoreType.ALIGNMENT.value: alignment_score,
                    ScoreType.COVERAGE.value: coverage_score,
                }
                self.score = min(alignment_score, coverage_score)
                self.reason = self._generate_reason()
                self.success = self.score >= self.threshold
                self.verbose_logs = construct_verbose_logs(
                    self,
                    steps=[
                        f"Truths (limit={self.truths_extraction_limit}):\n{prettify_list(self.truths)}",
                        f"Claims:\n{prettify_list(self.claims)}",
                        f"Assessment Questions:\n{prettify_list(self.assessment_questions)}",
                        f"Coverage Verdicts:\n{prettify_list(self.coverage_verdicts)}",
                        f"Alignment Verdicts:\n{prettify_list(self.alignment_verdicts)}",
                        f"Score: {self.score}\nReason: {self.reason}",
                    ],
                )

            return self.score

    async def a_measure(
        self,
        test_case: LLMTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
    ) -> float:

        check_llm_test_case_params(test_case, self._required_params, self)

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self,
            async_mode=True,
            _show_indicator=_show_indicator,
            _in_component=_in_component,
        ):
            self.truths, self.claims = await asyncio.gather(
                self._a_generate_truths(test_case.input),
                self._a_generate_claims(test_case.actual_output),
            )
            (
                self.coverage_verdicts,
                self.alignment_verdicts,
            ) = await asyncio.gather(
                self._a_generate_coverage_verdicts(test_case),
                self._a_generate_alignment_verdicts(),
            )
            alignment_score = self._calculate_score(ScoreType.ALIGNMENT)
            coverage_score = self._calculate_score(ScoreType.COVERAGE)
            self.score_breakdown = {
                ScoreType.ALIGNMENT.value: alignment_score,
                ScoreType.COVERAGE.value: coverage_score,
            }
            self.score = min(alignment_score, coverage_score)
            self.reason = await self._a_generate_reason()
            self.success = self.score >= self.threshold
            self.verbose_logs = construct_verbose_logs(
                self,
                steps=[
                    f"Truths (limit={self.truths_extraction_limit}):\n{prettify_list(self.truths)}",
                    f"Claims:\n{prettify_list(self.claims)}",
                    f"Assessment Questions:\n{prettify_list(self.assessment_questions)}",
                    f"Coverage Verdicts:\n{prettify_list(self.coverage_verdicts)}",
                    f"Alignment Verdicts:\n{prettify_list(self.alignment_verdicts)}",
                    f"Score: {self.score}\nReason: {self.reason}",
                ],
            )

            return self.score

    async def _a_generate_reason(self) -> str:
        if self.include_reason is False:
            return None

        contradictions = []
        redundancies = []
        for verdict in self.alignment_verdicts:
            if verdict.verdict.strip().lower() == "no":
                contradictions.append(verdict.reason)
            elif verdict.verdict.strip().lower() == "idk":
                redundancies.append(verdict.reason)

        questions = []
        if self.coverage_verdicts:
            for verdict in self.coverage_verdicts:
                if (
                    verdict.original_verdict.strip().lower() == "yes"
                    and verdict.summary_verdict.strip().lower() == "no"
                ):
                    questions.append(verdict.question)

        prompt: dict = SummarizationTemplate.generate_reason(
            contradictions=contradictions,
            redundancies=redundancies,
            questions=questions,
            score=format(self.score, ".2f"),
        )

        if len(questions) > 0:
            prompt += f"""Questions the original text can answer but not the summary:
{questions}

"""
        prompt += """JSON:
"""

        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Reason)
            self.evaluation_cost += cost
            return res.reason
        else:
            try:
                res: Reason = await self.model.a_generate(prompt, schema=Reason)
                return res.reason
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    def _generate_reason(self) -> str:
        if self.include_reason is False:
            return None

        contradictions = []
        redundancies = []
        for verdict in self.alignment_verdicts:
            if verdict.verdict.strip().lower() == "no":
                contradictions.append(verdict.reason)
            elif verdict.verdict.strip().lower() == "idk":
                redundancies.append(verdict.reason)

        questions = []
        if self.coverage_verdicts:
            for verdict in self.coverage_verdicts:
                if (
                    verdict.original_verdict.strip().lower() == "yes"
                    and verdict.summary_verdict.strip().lower() == "no"
                ):
                    questions.append(verdict.question)

        prompt: dict = SummarizationTemplate.generate_reason(
            contradictions=contradictions,
            redundancies=redundancies,
            questions=questions,
            score=format(self.score, ".2f"),
        )

        if len(questions) > 0:
            prompt += f"""Questions the original text can answer but not the summary:
{questions}

"""
        prompt += """JSON:
"""

        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Reason)
            self.evaluation_cost += cost
            return res.reason
        else:
            try:
                res: Reason = self.model.generate(prompt, schema=Reason)
                return res.reason
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    def _calculate_score(self, score_type: ScoreType) -> float:
        if score_type == ScoreType.ALIGNMENT:
            total = len(self.alignment_verdicts)
            if total == 0:
                return 0
            faithfulness_count = 0
            for verdict in self.alignment_verdicts:
                # Different from the faithfulness score, this
                # penalizes 'idk' (full of fluff) summaries
                if verdict.verdict.strip().lower() == "yes":
                    faithfulness_count += 1

            score = faithfulness_count / total

        else:
            if self.assessment_questions is None:
                return 1
            total = 0
            coverage_count = 0
            for verdict in self.coverage_verdicts:
                if verdict.original_verdict.strip().lower() == "yes":
                    total += 1
                    if verdict.summary_verdict.strip().lower() == "yes":
                        coverage_count += 1

            if total == 0:
                return 0

            score = coverage_count / total

        return 0 if self.strict_mode and score < self.threshold else score

    async def _a_generate_answers(self, text: str) -> List[str]:
        prompt = SummarizationTemplate.generate_answers(
            questions=self.assessment_questions, text=text
        )
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Answers)
            self.evaluation_cost += cost
            return res.answers
        else:
            try:
                res: Answers = await self.model.a_generate(
                    prompt, schema=Answers
                )
                return res.answers
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["answers"]

    def _generate_answers(self, text: str) -> List[str]:
        prompt = SummarizationTemplate.generate_answers(
            questions=self.assessment_questions, text=text
        )
        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Answers)
            self.evaluation_cost += cost
            return res.answers
        else:
            try:
                res: Answers = self.model.generate(prompt, schema=Answers)
                return res.answers
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["answers"]

    async def _a_generate_assessment_questions(self, text: str):
        prompt = SummarizationTemplate.generate_questions(text=text, n=self.n)
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Questions)
            self.evaluation_cost += cost
            return res.questions
        else:
            try:
                res: Questions = await self.model.a_generate(
                    prompt, schema=Questions
                )
                return res.questions
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["questions"]

    def _generate_assessment_questions(self, text: str):
        prompt = SummarizationTemplate.generate_questions(text=text, n=self.n)
        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Questions)
            self.evaluation_cost += cost
            return res.questions
        else:
            try:
                res: Questions = self.model.generate(prompt, schema=Questions)
                return res.questions
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["questions"]

    async def _a_generate_coverage_verdicts(
        self, test_case: LLMTestCase
    ) -> List[SummarizationCoverageVerdict]:
        if self.assessment_questions is None:
            self.assessment_questions = (
                await self._a_generate_assessment_questions(test_case.input)
            )

        tasks = [
            self._a_generate_answers(test_case.input),
            self._a_generate_answers(test_case.actual_output),
        ]
        results = await asyncio.gather(*tasks)
        original_answers = results[0]
        summary_answers = results[1]

        if len(original_answers) != len(summary_answers):
            raise ValueError("Number of verdicts generated does not equal.")

        coverage_verdicts: List[SummarizationCoverageVerdict] = []
        for i in range(len(original_answers)):
            coverage_verdicts.append(
                SummarizationCoverageVerdict(
                    summary_verdict=summary_answers[i],
                    original_verdict=original_answers[i],
                    question=self.assessment_questions[i],
                )
            )
        return coverage_verdicts

    def _generate_coverage_verdicts(
        self, test_case: LLMTestCase
    ) -> List[SummarizationCoverageVerdict]:
        if self.assessment_questions is None:
            self.assessment_questions = self._generate_assessment_questions(
                test_case.input
            )

        original_answers = self._generate_answers(test_case.input)
        summary_answers = self._generate_answers(test_case.actual_output)

        if len(original_answers) != len(summary_answers):
            raise ValueError("Number of verdicts generated does not equal.")

        coverage_verdicts: List[SummarizationCoverageVerdict] = []
        for i in range(len(original_answers)):
            coverage_verdicts.append(
                SummarizationCoverageVerdict(
                    summary_verdict=summary_answers[i],
                    original_verdict=original_answers[i],
                    question=self.assessment_questions[i],
                )
            )

        return coverage_verdicts

    async def _a_generate_alignment_verdicts(
        self,
    ) -> List[SummarizationAlignmentVerdict]:
        if len(self.claims) == 0:
            return []

        verdicts: List[SummarizationAlignmentVerdict] = []
        prompt = SummarizationTemplate.generate_alignment_verdicts(
            summary_claims=self.claims, original_text="\n\n".join(self.truths)
        )
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Verdicts)
            self.evaluation_cost += cost
            verdicts = [item for item in res.verdicts]
            return verdicts
        else:
            try:
                res: Verdicts = await self.model.a_generate(
                    prompt, schema=Verdicts
                )
                verdicts = [item for item in res.verdicts]
                return verdicts
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                verdicts = [
                    SummarizationAlignmentVerdict(**item)
                    for item in data["verdicts"]
                ]
                return verdicts

    def _generate_alignment_verdicts(
        self,
    ) -> List[SummarizationAlignmentVerdict]:
        if len(self.claims) == 0:
            return []

        verdicts: List[SummarizationAlignmentVerdict] = []
        prompt = SummarizationTemplate.generate_alignment_verdicts(
            summary_claims=self.claims, original_text="\n\n".join(self.truths)
        )
        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Verdicts)
            self.evaluation_cost += cost
            verdicts = [item for item in res.verdicts]
            return verdicts
        else:
            try:
                res: Verdicts = self.model.generate(prompt, schema=Verdicts)
                verdicts = [item for item in res.verdicts]
                return verdicts
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                verdicts = [
                    SummarizationAlignmentVerdict(**item)
                    for item in data["verdicts"]
                ]
                return verdicts

    async def _a_generate_truths(self, text: str) -> List[str]:
        # Borrow faithfulness template
        prompt = FaithfulnessTemplate.generate_truths(
            retrieval_context=text,
            extraction_limit=self.truths_extraction_limit,
        )
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Truths)
            self.evaluation_cost += cost
            return res.truths
        else:
            try:
                res: Truths = await self.model.a_generate(prompt, schema=Truths)
                return res.truths
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["truths"]

    async def _a_generate_claims(self, text: str) -> List[str]:
        # Borrow faithfulness template
        prompt = FaithfulnessTemplate.generate_claims(actual_output=text)
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Claims)
            self.evaluation_cost += cost
            return res.claims
        else:
            try:
                res: Claims = await self.model.a_generate(prompt, schema=Claims)
                return res.claims
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["claims"]

    def _generate_truths(self, text: str) -> List[str]:
        # Borrow faithfulness template
        prompt = FaithfulnessTemplate.generate_truths(
            retrieval_context=text,
            extraction_limit=self.truths_extraction_limit,
        )
        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Truths)
            self.evaluation_cost += cost
            return res.truths
        else:
            try:
                res: Truths = self.model.generate(prompt, schema=Truths)
                return res.truths
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["truths"]

    def _generate_claims(self, text: str) -> List[str]:
        # Borrow faithfulness template
        prompt = FaithfulnessTemplate.generate_claims(actual_output=text)
        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Claims)
            self.evaluation_cost += cost
            return res.claims
        else:
            try:
                res: Claims = self.model.generate(prompt, schema=Claims)
                return res.claims
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["claims"]

    def is_successful(self) -> bool:
        if self.error is not None:
            self.success = False
        else:
            try:
                self.success = self.score >= self.threshold
            except:
                self.success = False
        return self.success

    @property
    def __name__(self):
        return "Summarization"

File: typescript/src/metrics/summarization/summarization.ts
import { BaseMetric } from "../base-metrics";
import { LLMTestCase, SingleTurnParams } from "../../test-case";
import { DeepEvalBaseLLM } from "../../models";
import { resolveTemplate } from "../../templates";
import {
  initializeModel,
  generateWithSchema,
  checkSingleTurnParams,
  constructVerboseLogs,
  prettifyList,
} from "../utils";
import { TruthsSchema, ClaimsSchema } from "../faithfulness/schema";
import {
  VerdictsSchema,
  QuestionsSchema,
  AnswersSchema,
  SummarizationScoreReasonSchema,
  type SummarizationAlignmentVerdict,
  type SummarizationCoverageVerdict,
} from "./schema";

const TEMPLATE_CLASS = "SummarizationMetric";

// Borrowed from faithfulness (Python imports `_faithfulness_truths_limit_phrase`).
function truthsLimitPhrase(limit?: number): string {
  if (limit == null) return " FACTUAL, undisputed truths";
  if (limit === 1) return " the single most important FACTUAL, undisputed truth";
  return ` the ${limit} most important FACTUAL, undisputed truths per document`;
}

/** Python `repr()` of a list of strings, for the manually-appended reason prompt. */
function pyListRepr(items: string[]): string {
  return (
    "[" +
    items
      .map((s) => "'" + s.replace(/\\/g, "\\\\").replace(/'/g, "\\'") + "'")
      .join(", ") +
    "]"
  );
}

export interface SummarizationMetricOptions {
  threshold?: number;
  /** Number of assessment questions to generate (when none are supplied). */
  n?: number;
  model?: DeepEvalBaseLLM | string;
  /** Pre-supplied yes/no assessment questions (skips question generation). */
  assessmentQuestions?: string[];
  includeReason?: boolean;
  strictMode?: boolean;
  verboseMode?: boolean;
  showIndicator?: boolean;
  truthsExtractionLimit?: number;
}

/**
 * Summarization — is the summary (`actualOutput`) both faithful to and covers
 * the original text (`input`)? Combines an **alignment** score (summary claims
 * vs. original truths) and a **coverage** score (assessment Q&A on both), and
 * takes `score = min(alignment, coverage)`. **Higher is better**.
 */
export class SummarizationMetric extends BaseMetric {
  truths: string[] = [];
  claims: string[] = [];
  coverageVerdicts: SummarizationCoverageVerdict[] = [];
  alignmentVerdicts: SummarizationAlignmentVerdict[] = [];
  assessmentQuestions?: string[];
  private readonly n: number;
  private readonly truthsExtractionLimit?: number;

  constructor(options: SummarizationMetricOptions = {}) {
    const strictMode = options.strictMode ?? false;
    super(strictMode ? 1 : (options.threshold ?? 0.5), {
      strictMode,
      verboseMode: options.verboseMode,
      includeReason: options.includeReason ?? true,
      showIndicator: options.showIndicator,
    });
    this.requiredParams = [
      SingleTurnParams.INPUT,
      SingleTurnParams.ACTUAL_OUTPUT,
    ];
    this.n = options.n ?? 5;
    this.assessmentQuestions =
      options.assessmentQuestions && options.assessmentQuestions.length > 0
        ? options.assessmentQuestions
        : undefined;
    this.truthsExtractionLimit =
      options.truthsExtractionLimit != null
        ? Math.max(options.truthsExtractionLimit, 0)
        : undefined;
    const { model, usingNativeModel } = initializeModel(options.model);
    this.model = model;
    this.usingNativeModel = usingNativeModel;
    this.evaluationModel = this.model.getModelName();
  }

  async measure(testCase: LLMTestCase): Promise<number> {
    this.error = undefined;
    await this.startProgress();
    try {
      checkSingleTurnParams(testCase, this.requiredParams, this);
      this.evaluationCost = this.usingNativeModel ? 0 : undefined;

      [this.truths, this.claims] = await Promise.all([
        this.generateTruths(testCase.input),
        this.generateClaims(testCase.actualOutput),
      ]);
      [this.coverageVerdicts, this.alignmentVerdicts] = await Promise.all([
        this.generateCoverageVerdicts(testCase),
        this.generateAlignmentVerdicts(),
      ]);

      const alignmentScore = this.calculateAlignmentScore();
      const coverageScore = this.calculateCoverageScore();
      this.scoreBreakdown = {
        Alignment: alignmentScore,
        Coverage: coverageScore,
      };
      this.score = Math.min(alignmentScore, coverageScore);
      this.reason = await this.generateReason();
      this.success = this.score >= this.threshold;

      this.verboseLogs = constructVerboseLogs(this, [
        `Truths (limit=${this.truthsExtractionLimit}):\n${prettifyList(this.truths)}`,
        `Claims:\n${prettifyList(this.claims)}`,
        `Assessment Questions:\n${prettifyList(this.assessmentQuestions ?? [])}`,
        `Coverage Verdicts:\n${prettifyList(this.coverageVerdicts)}`,
        `Alignment Verdicts:\n${prettifyList(this.alignmentVerdicts)}`,
        `Score: ${this.score}\nReason: ${this.reason}`,
      ]);
      return this.score;
    } finally {
      this.stopProgress();
    }
  }

  // --- truths/claims (borrow Faithfulness templates) ---

  private async generateTruths(text: string): Promise<string[]> {
    const prompt = resolveTemplate("metrics", "FaithfulnessMetric", "generate_truths", {
      retrieval_context: text,
      limit: truthsLimitPhrase(this.truthsExtractionLimit),
      multimodal_instruction: "",
    });
    const { truths } = await generateWithSchema(this, prompt, TruthsSchema);
    return truths;
  }

  private async generateClaims(text: string): Promise<string[]> {
    const prompt = resolveTemplate("metrics", "FaithfulnessMetric", "generate_claims", {
      actual_output: text,
      multimodal_instruction: "",
    });
    const { claims } = await generateWithSchema(this, prompt, ClaimsSchema);
    return claims;
  }

  // --- alignment ---

  private async generateAlignmentVerdicts(): Promise<
    SummarizationAlignmentVerdict[]
  > {
    if (this.claims.length === 0) return [];
    const prompt = resolveTemplate("metrics", 
      TEMPLATE_CLASS,
      "generate_alignment_verdicts",
      {
        summary_claims: this.claims,
        original_text: this.truths.join("\n\n"),
      },
    );
    const { verdicts } = await generateWithSchema(this, prompt, VerdictsSchema);
    return verdicts;
  }

  // --- coverage ---

  private async generateCoverageVerdicts(
    testCase: LLMTestCase,
  ): Promise<SummarizationCoverageVerdict[]> {
    if (this.assessmentQuestions == null) {
      this.assessmentQuestions = await this.generateAssessmentQuestions(
        testCase.input,
      );
    }
    const [originalAnswers, summaryAnswers] = await Promise.all([
      this.generateAnswers(testCase.input),
      this.generateAnswers(testCase.actualOutput),
    ]);
    if (originalAnswers.length !== summaryAnswers.length) {
      throw new Error("Number of verdicts generated does not equal.");
    }
    const coverage: SummarizationCoverageVerdict[] = [];
    for (let i = 0; i < originalAnswers.length; i++) {
      coverage.push({
        summary_verdict: summaryAnswers[i],
        original_verdict: originalAnswers[i],
        question: this.assessmentQuestions[i],
      });
    }
    return coverage;
  }

  private async generateAssessmentQuestions(text: string): Promise<string[]> {
    const prompt = resolveTemplate("metrics", TEMPLATE_CLASS, "generate_questions", {
      text,
      n: this.n,
    });
    const { questions } = await generateWithSchema(
      this,
      prompt,
      QuestionsSchema,
    );
    return questions;
  }

  private async generateAnswers(text: string): Promise<string[]> {
    const prompt = resolveTemplate("metrics", TEMPLATE_CLASS, "generate_answers", {
      questions: this.assessmentQuestions,
      text,
    });
    const { answers } = await generateWithSchema(this, prompt, AnswersSchema);
    return answers;
  }

  // --- reason ---

  private async generateReason(): Promise<string | undefined> {
    if (!this.includeReason) return undefined;
    const contradictions: (string | null | undefined)[] = [];
    const redundancies: (string | null | undefined)[] = [];
    for (const v of this.alignmentVerdicts) {
      const vd = v.verdict.trim().toLowerCase();
      if (vd === "no") contradictions.push(v.reason);
      else if (vd === "idk") redundancies.push(v.reason);
    }
    const questions: string[] = [];
    for (const v of this.coverageVerdicts) {
      if (
        v.original_verdict.trim().toLowerCase() === "yes" &&
        v.summary_verdict.trim().toLowerCase() === "no"
      ) {
        questions.push(v.question);
      }
    }

    let prompt = resolveTemplate("metrics", TEMPLATE_CLASS, "generate_reason", {
      contradictions,
      redundancies,
      questions,
      score: (this.score ?? 0).toFixed(2),
    });
    if (questions.length > 0) {
      prompt += `Questions the original text can answer but not the summary:\n${pyListRepr(questions)}\n\n`;
    }
    prompt += "JSON:\n";

    const { reason } = await generateWithSchema(
      this,
      prompt,
      SummarizationScoreReasonSchema,
    );
    return reason;
  }

  // --- scoring (alignment penalizes idk; coverage = answerable-by-both / answerable-by-original) ---

  private calculateAlignmentScore(): number {
    const total = this.alignmentVerdicts.length;
    if (total === 0) return 0;
    const faithful = this.alignmentVerdicts.filter(
      (v) => v.verdict.trim().toLowerCase() === "yes",
    ).length;
    const score = faithful / total;
    return this.strictMode && score < this.threshold ? 0 : score;
  }

  private calculateCoverageScore(): number {
    if (this.assessmentQuestions == null) return 1;
    let total = 0;
    let coverage = 0;
    for (const v of this.coverageVerdicts) {
      if (v.original_verdict.trim().toLowerCase() === "yes") {
        total++;
        if (v.summary_verdict.trim().toLowerCase() === "yes") coverage++;
      }
    }
    if (total === 0) return 0;
    const score = coverage / total;
    return this.strictMode && score < this.threshold ? 0 : score;
  }

  isSuccessful(): boolean {
    const ok = this.error == null && (this.score ?? 0) >= this.threshold;
    this.success = ok;
    return ok;
  }

  get name(): string {
    return "Summarization";
  }
}
